:books: Docs sur clustering & déduplication (#1394)
incubateur-ademe/quefairedemesobjets: docs/reference/data/clustering_deduplication.md

# Clustering & Déduplication

- 🗺️ **Contexte**: on ingère de la donnée d’acteurs de l’économie circulaire de [plusieurs sources](https://data.ademe.fr/)
- 🔴 **Problème**: certains acteurs sont en doublon, polluant l'application utilisteur
- 🟢 **Solution**: faire un travail pour réduire les doublons et améliorer la cohérence des données
- 🤔 **Comment**: voir ci-dessous

## ⚠️ Limitations connues

- **On ne séparent pas les enfants d’un cluster existant**: choix de notre part, pour l’instant on veut uniquement
  ajouter des nouveaux enfants. Pour changer il faudra [CLUSTERING & DEDUP: re-clusteriser les enfants existants](https://www.notion.so/CLUSTERING-DEDUP-re-clusteriser-les-enfants-existants-1956523d57d7801abe3bc132ec7e33b0?pvs=21). Pour l’instant:
  - 🟢 **avantage**: pas de risque d’endommager des clusters existants
  - 🟠 **inconvénient**: pas d’opportunité de re-clustering les mauvais cluster existants
 
- **Pas de re-clustering enfants = pas de contexte enfant**: conséquence du dessus, les enfants n’ont pas leur donnée récupérée ni

>   normalisée, ce qui peut poser des problèmes de contexte, donc on à fait https://github.com/incubateur-ademe/quefairedemesobjets/pull/1379 en attendant

Reviewer: ```suggestion
  normalisée, ce qui peut poser des problèmes de contexte, donc on a [exclus les enfants existants du contexte](https://github.com/incubateur-ademe/quefairedemesobjets/pull/1379) en attendant
```

## 📜 Définition

> - 📦 **Clustering:** fait de **regrouper** des ****acteurs via leur **similarité**

Reviewer: ```suggestion
- 📦 **Clustering:** fait de **regrouper** des acteurs via leur **similarité**
```

- 1️⃣ **Déduplication**: convertir un **cluster** en **1 seul acteur** pour ne **plus avoir de doublons**
- 🎏 **Etat final idéal d’un acteur**: on se concentre sur les acteurs statut=ACTIF

> <table><thead>
>   <tr>
>     <th>Etat 🇫🇷</th>
>     <th>Code 🇬🇧</th>
>     <th>Définition</th>
>     <th>Correspond à une source en particulier</th>
>     <th>Visible sur la carte</th>
>   </tr></thead>
> <tbody>
>   <tr>
>     <td>parent(s)</td>
>     <td>parent(s)</td>
>     <td>acteur qui a 1 ou plusieurs enfants rattachés à lui</td>
>     <td>🟠 NON<br>on créé un parent “artificiel” pour que celui-ci vive de manière détachée de ses enfants = plus robuste à travers le temps, pas besoin de changer le parent à chaque fois que les enfants changent<br></td>
>     <td>🟢 OUI</td>
>   </tr>
>   <tr>
>     <td>enfant(s)</td>
>     <td>child / children</td>
>     <td>acteur rattaché à 1 parent (1 est le maximum)</td>
>     <td>🟢 OUI<br>C’est de cette source que vient l’acteur<br></td>
>     <td>🟠 NON<br>C’est le parent qui est affiché<br></td>
>   </tr>
>   <tr>
>     <td>orphelin(s)</td>
>     <td>orphan(s)</td>
>     <td>acteur rattaché à 0 parent</td>
>     <td>🟢 OUI<br>C’est de cette source que vient l’acteur<br></td>
>     <td>🟢 OUI</td>
>   </tr>
> </tbody></table>

Reviewer: syntax MD ?

```suggestion
| Etat 🇫🇷 | Code 🇬🇧 | Définition | Correspond à une source en particulier | Visible sur la carte |
|---------|----------|------------|------------------------------------|--------------------|
| parent(s) | parent(s) | acteur qui a 1 ou plusieurs enfants rattachés à lui | 🟠 NON <br> on créé un parent “artificiel” pour que celui-ci vive de manière détachée de ses enfants = plus robuste à travers le temps, pas besoin de changer le parent à chaque fois que les enfants changent | 🟢 OUI |
| enfant(s) | child / children | acteur rattaché à 1 parent (1 est le maximum) | 🟢 OUI <br> C’est de cette source que vient l’acteur | 🟠 NON <br> C’est le parent qui est affiché |
| orphelin(s) | orphan(s) | acteur rattaché à 0 parent | 🟢 OUI | C’est de cette source que vient l’acteur | 🟠 NON | C’est le parent qui est affiché |
| orphelin(s) | orphan(s) | acteur rattaché à 0 parent | 🟢 OUI <br> C’est de cette source que vient l’acteur | 🟠 NON <br> C’est le parent qui est affiché |
```

Author: Trop chian à maintenir, dès qu'on veut faire un petit edit de la table faut repasser par un formatteur markdown

## ➡️ Transitions d’états: scénarios

> <table><thead>

Reviewer: idem pour le format en MD ?

Author: Pour l'instant je vais m'en passer :smile:

    <td>Enfant</td>
    <td>Enfant</td>
    <td>🟰 Pointe déjà vers nouveau parent → rien à faire</td>
    <td>Aucune</td>
    <td>Aucune</td>
  </tr>
  <tr>
    <td><a href="/incubateur-ademe/quefairedemesobjets/blob/main/data/models/changes/acteur_update_parent_id.py">acteur_update_parent_id</a></td>
    <td>Enfant</td>
    <td>Enfant</td>
    <td>🔀 Pointait vers un parent qui n’a pas été choisi → à pointer vers nouveau parent</td>
    <td> 🔀 Mettre à jour parent_id pour pointer vers nouveau parent</td>
    <td>Aucune</td>
  </tr>
  <tr>
    <td><a href="/incubateur-ademe/quefairedemesobjets/blob/main/data/models/changes/acteur_update_parent_id.py">acteur_update_parent_id</a></td>
    <td>Orphelin</td>
    <td>Enfant</td>
    <td>🔀 à pointer vers un parent</td>
    <td> 🔀 Mettre à jour parent_id pour pointer vers nouveau parent</td>
    <td>🛑 Devrait disparaitre de displayed</td>
  </tr>
  <tr>
    <td></td>
    <td>Orphelin</td>
    <td>Orphelin</td>
    <td>Ne fais toujours pas parti d’un cluster (pas de changement)</td>
    <td>Aucune</td>
    <td>Aucune</td>
  </tr>
</tbody></table>

## 🔀 Schéma

> ```mermaid

Reviewer: Cristal clean :)
